app/controllers/KairosDbProxy.java

```diff
@@ -0,0 +1,358 @@
+     *
+     * @return Proxied tagNames response.
+     */
+    public CompletionStage<Result> tagNames() {
+        return proxy();
+    }
+
+    /**
+     * Proxied tagValues call.
+     *
+     * @return Proxied tagValues response.
+     */
+    public CompletionStage<Result> tagValues() {
+        return proxy();
+    }
+
+    /**
+     * Proxied queryTags call.
+     *
+     * @return Proxied queryTags response.
+     */
+    public CompletionStage<Result> queryTags() {
+        return proxy();
+    }
+
+    /**
+     * Proxied queryMetrics call.
+     *
+     * @return Proxied queryMetrics response.
+     */
+    public CompletionStage<Result> queryMetrics() {
+        return proxy();
+    }
+
+    /**
+     * Proxied version call.
+     *
+     * @return Proxied version response.
+     */
+    public CompletionStage<Result> version() {
+        return proxy();
+    }
+
+    /**
+     * Caching metricNames call.
+     *
+     * @param containing simple string match filter for metric names
+     * @return Cached metric names, filtered by the query string.
+     */
+    public CompletionStage<Result> metricNames(final String containing) {
+        final List<String> metrics = _cache.getIfPresent(METRICS_KEY);
+
+        final CompletionStage<List<String>> response;
+        if (metrics != null) {
+            response = CompletableFuture.completedFuture(metrics);
+        } else {
+            // TODO(brandon): Investigate refreshing eagerly or in the background
+            // Refresh the cache
+            final CompletionStage<List<String>> queryResponse = _kairosDbClient.queryMetricNames()
+                    .thenApply(KairosMetricNamesQueryResponse::getResults)
+                    .thenApply(list -> {
+                        _cache.put(METRICS_KEY, list);
+                        _metricsList.set(list);
+                        return list;
+                    });
+
+            final List<String> metricsList = _metricsList.get();
+            if (metricsList != null) {
+                response = CompletableFuture.completedFuture(metricsList);
+            } else {
+                response = queryResponse;
+            }
+        }
+
+        return response
+                .thenApply(list -> filterMetricNames(list, containing))
+                .thenApply(list -> new KairosMetricNamesQueryResponse.Builder().setResults(list).build())
+                .<JsonNode>thenApply(_mapper::valueToTree)
+                .thenApply(Results::ok);
+    }
+
+    private static ImmutableList<String> filterMetricNames(final List<String> metricNames, final String containing) {
+        if (containing == null || containing.isEmpty()) {
+            return ImmutableList.of();
+        }
+
+        final String lowerContaining = containing.toLowerCase(Locale.ENGLISH);
+        final Predicate<String> containsPredicate = s -> s.toLowerCase().contains(lowerContaining);
+
+        return metricNames.stream()
+                .filter(IS_PT1M.negate())
```

Comment:
I don't understand this restriction. Our aggregates can be for any configurable time period.

Reply:
Histograms no longer have "PT1M/" as the prefix.  This filters out everything that's not a histogram.  The future is now ( :